File: multiobjective.py
# This Python file uses the following encoding: utf-8
import sys

from PySide6.QtWidgets import QApplication, QMainWindow
from Optimisers.MultiObjective import EigenvectorMethod
from Optimisers.MultiObjective import FuzzyLogic
from Optimisers.MultiObjective import ParetoDominance
from Optimisers.MultiObjective import ParetoDominanceTable

# Important:
# You need to run the following command to generate the ui_form.py file
#     pyside6-uic form.ui -o ui_form.py, or
#     pyside2-uic form.ui -o ui_form.py

from ui_multiobjective import Ui_wMultiObjective


class MultiObjectiveWindow(QMainWindow):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.ui = Ui_wMultiObjective()
        self.ui.setupUi(self)

        self.ui.btnCalculate.clicked.connect(self.btnCalculate_clicked)
        self.ui.btnBack.clicked.connect(self.btnBack_clicked)
        print("Initialised")

    def btnCalculate_clicked(self):
        print("Clicked")
        ShowPlot = False
        if self.ui.cbPlot.isChecked():
            ShowPlot = True

        Functions = str(self.ui.txtFunctions.toPlainText()).split("\n")
        print(Functions)

        x_values = []
        thresholds = []
        Left, Right = [], []

        if self.ui.txtLeft.toPlainText().strip():
            Left = self.ui.txtLeft.toPlainText().split("\n")

        if self.ui.txtRight.toPlainText().strip():
            Right = self.ui.txtRight.toPlainText().split("\n")

        if Right and Left:
            for i in range(len(Left)):
                thresholds.append((float(Left[i]), float(Right[i])))
        elif Left:
            for val in Left:
                val = val.strip()
                if val:
                    x_values.append(float(val))

        print(thresholds)

        samples = []
        for i in range (1, len(x_values) + 1):
            samples.append(i)



        if self.ui.rbEigenvectors.isChecked():
            print("Starting eigenvectors")
            result = EigenvectorMethod.eigenvector_weights(Functions)
            print("Finished eigenvectors")
        elif self.ui.rbFuzzyLogic.isChecked():
            print("Starting fuzzy logic")
            result = FuzzyLogic.fuzzy(Functions, thresholds)
            print("Finished fuzzy logic")
        elif self.ui.rbPareto.isChecked():
            print("Starting Pareto")
            result = ParetoDominance.pareto_front(Functions, ShowPlot)
            print("Finished Pareto")
        elif self.ui.rbParetoTable.isChecked():
            print("Starting Pareto Table")
            result = ParetoDominanceTable.pareto_analysis(Functions, x_values, samples, ShowPlot)
            print("Finished Pareto Table")

        self.ui.txtOutput.setText(str(result))

    def btnBack_clicked(self):
        from menu import MainWindow
        print("Back")
        self.hide()
        self.ui_menu = MainWindow()
        self.ui_menu.show()


if __name__ == "__main__":
    app = QApplication(sys.argv)
    widget = MultiObjectiveWindow()
    widget.show()
    sys.exit(app.exec())

File: multiobjective.cpp
// multiobjective.cpp
#include "multiobjective.h"
#include "menu.h"
#include "python_utils.h"
#include <QDebug>

MultiObjectiveWindow::MultiObjectiveWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::wMultiObjective)
{
    ui->setupUi(this);
    connect(ui->btnCalculate, &QPushButton::clicked,
            this, &MultiObjectiveWindow::btnCalculate_clicked);
    connect(ui->btnBack, &QPushButton::clicked,
            this, &MultiObjectiveWindow::btnBack_clicked);
    qDebug() << "MultiObjectiveWindow initialized";
}

MultiObjectiveWindow::~MultiObjectiveWindow()
{
    delete ui;
}

void MultiObjectiveWindow::btnCalculate_clicked()
{
    qDebug() << "Calculate clicked";
    bool showPlot = ui->cbPlot->isChecked();

    // Gather functions
    QStringList functions = ui->txtFunctions
                                ->toPlainText()
                                .split('\n', Qt::SkipEmptyParts);

    // Gather thresholds
    QStringList left  = ui->txtLeft
                           ->toPlainText()
                           .split('\n', Qt::SkipEmptyParts);
    QStringList right = ui->txtRight
                            ->toPlainText()
                            .split('\n', Qt::SkipEmptyParts);

    QStringList args;
    QString module, function;

    if (ui->rbEigenvectors->isChecked()) {
        module   = "Optimisers.MultiObjective.EigenvectorMethod";
        function = "eigenvector_weights";
        // single arg: functions
        args << functions.join(";");
    }
    else if (ui->rbFuzzyLogic->isChecked()) {
        module   = "Optimisers.MultiObjective.FuzzyLogic";
        function = "fuzzy";
        // args: functions, thresholds as "l1,r1;l2,r2;..."
        QStringList thr;
        for (int i = 0; i < qMin(left.size(), right.size()); ++i)
            thr << left[i] + "," + right[i];
        args << functions.join(";")
             << thr.join(";");
    }
    else if (ui->rbPareto->isChecked()) {
        module   = "Optimisers.MultiObjective.ParetoDominance";
        function = "pareto_front";
        args << functions.join(";")
             << (showPlot ? "1" : "0");
    }
    else if (ui->rbParetoTable->isChecked()) {
        module   = "Optimisers.MultiObjective.ParetoDominanceTable";
        function = "pareto_analysis";
        // build x_values from left only
        QStringList xvals = left;
        // samples = 1..xvals.size()
        QStringList samples;
        for (int i = 1; i <= xvals.size(); ++i)
            samples << QString::number(i);
        args << functions.join(";")
             << xvals.join(";")
             << samples.join(";")
             << (showPlot ? "1" : "0");
    }

    // Call into Python
    QString result = callPythonFunction(module, function, args);
    qDebug() << "Python returned:" << result;
    ui->txtOutput->setText(result);
}

void MultiObjectiveWindow::btnBack_clicked()
{
    qDebug() << "Back clicked";
    hide();
    Menu *m = new Menu(this);
    m->show();
}
